### PR TITLE
Make realm dynamic

### PR DESCRIPTION
**Pull request by mcollina**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -42,25 +44,28 @@ function basicPlugin (fastify, opts, next) {
...
+    switch (typeof realm) {
+      case 'undefined':
+        return 'Basic'
+      case 'boolean':
+        return 'Basic'
+      case 'string':
+        return `Basic realm="${realm}"` 
+      case 'function':
+        return function (req) {
+          return `Basic realm="${realm(req)}"`
```
Comment (climba03003) [inline index.js:63]: Should we support async function for `realm` ?

Comment (mcollina) [reply index.js]: I'm not convinced of the API either. I think it might be better to have `validate` provide the realm.

**Comment by mcollina** [comment]
@climba03003 PTAL, now it is simpler and async.

**Review by climba03003 — APPROVED** [review]
LGTM.

File: index.js [hunk 2]
```diff
@@ -28,39 +23,51 @@ function basicPlugin (fastify, opts, next) {
...
+      }
+      if (err) {
         // We set the status code to be 401 if it is not set
         if (!err.statusCode) {
           err.statusCode = 401
         }
         next(err)
       } else {
+        const header = realm ? formatRealm(realm) : authenticateHeader
+        reply.header('WWW-Authenticate', header)
```
Comment (Eomm) [inline index.js:40]: Should not this header be added on 401 only?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication

Comment (mcollina) [reply index.js:40]: You are right, I made a mess. I'll revert and restore the double callback. Sigh.

**Pull request merged**